app/src/main/java/org/breezyweather/sources/CommonConverter.kt

```diff
@@ -673,48 +678,190 @@ private fun completeHalfDayFromHourlyList(
+                    in 0.0..1.0 -> WeatherCode.FOG
+                    else -> null
+                }
+
+                // find closest hour with correct weather code
+                hour = sorted.firstOrNull { it.weatherCode == weatherCode }
+                if (hour == null) {
+                    // find the most mysterious hour and set a custom weather code
+                    hour = sorted.filter { it.visibility != null }.minByOrNull { it.visibility!! }
+                    customWeatherCode = weatherCode
+                }
+            } else if (maxWind?.speed != null && maxWind.speed > 10) {
+                // Average winds > 10m/s, its windy
+                val weatherCode = WeatherCode.WIND
+                // find closest hour with correct weather code
+                hour = sorted.firstOrNull { it.weatherCode == weatherCode }
+                if (hour == null) {
+                    // find the closest windy hour and set a custom weather code
+                    hour = sorted.firstOrNull { it.wind?.speed != null && it.wind.speed > 10 }
+                    customWeatherCode = weatherCode
+                }
+            } else if (avgCloudCover != null) {
+                // It's not raining, its not windy, and its not mysterious. Just cloudy
+                val offsets = arrayMapOf(
+                    WeatherCode.CLOUDY to 75.0..100.0,
+                    WeatherCode.PARTLY_CLOUDY to 37.5..75.0,
+                    WeatherCode.CLEAR to 0.0..37.5
+                )
+                val weatherCode = when (avgCloudCover.toDouble()) {
+                    in offsets[WeatherCode.CLOUDY]!! -> WeatherCode.CLOUDY
+                    in offsets[WeatherCode.PARTLY_CLOUDY]!! -> WeatherCode.PARTLY_CLOUDY
+                    in offsets[WeatherCode.CLEAR]!! -> WeatherCode.CLEAR
+                    else -> null
+                }
+
+                // find closest hour with correct weather code
+                hour = sorted.firstOrNull { it.weatherCode == weatherCode }
+                if (hour == null) {
+                    // find the closest cloudy hour and set a custom weather code
+                    hour = sorted.filter { it.cloudCover != null }
+                        .firstOrNull { it.cloudCover!!.toDouble() in offsets[weatherCode]!! }
+                    customWeatherCode = weatherCode
+                }
+            }
+        }
+
+        // fallback to ideal hour (12:00 for day and 24:00 for night)
+        hour = hour ?: sorted.firstOrNull { it.weatherCode != null } ?: sorted.first()
 
+        // prioritize custom weather code if set
         if (halfDayWeatherCode == null) {
-            halfDayWeatherCode = halfDayHourlyListWeather.weatherCode
+            halfDayWeatherCode = customWeatherCode ?: hour.weatherCode
         }
         if (halfDayWeatherPhase == null) {
-            halfDayWeatherPhase = halfDayHourlyListWeather.weatherText
+            halfDayWeatherPhase = hour.weatherText
```

Comment:
We have R.array.live_wallpaper_weather_kinds that can provide text for each weather code, we shouldn't take the closest matching hour text as it applies only for that hour, and not for the whole half day. Example: weather text could be "Snow for 25 min", or things very specific to this hour.

Reply:
We should probably add a getWeatherText in WeatherViewController object.

Reply:
Unfortunately, did not find a beautiful solution for this. I have created `WeatherViewController.getWeatherText(WeatherCode)` which throws a NotImplementedException() for now.

Reply:
If weatherText is missing, it is now autocompleted using R.array.live_wallpaper_weather_kinds (in WeatherViewController.getWeatherText)